### PR TITLE
fix(client): Make public path on client match dev server publicPath

### PR DESCRIPTION
**Pull request by Loonride**
- [x] This is a **bugfix**
- [ ] This is a **feature**
- [ ] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Yes

### Motivation / Use-Case

Fixes: https://github.com/webpack/webpack-dev-server/issues/1385

This creates an absolute URL based on provided `devServer.publicPath`, host, port, etc., then it sets `__webpack_public_path__` on the client to this absolute URL. This forces the client to request `hot-update.json` from the correct location.

This works because `hot-update.json` is requested from the webpack public path here: https://github.com/webpack/webpack/blob/master/lib/web/JsonpMainTemplate.runtime.js#L31 (`$require$.p` is public path).

The client must be informed of `publicPath`, so I pass it to the client using resource query along with `sockHost`, `sockPort`, `sockPath`. I did some work to improve tests and how these options are encoded/decoded through the resource query.

### Breaking Changes

Now, the `__webpack_public_path__` will be set to an absolute URL.

### CODE REVIEW

File: client-src/default/utils/updatePublicPath.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const getUrlParts = require('./getUrlParts');
+
+function updatePublicPath(resourceQuery) {
+  const urlParts = getUrlParts(resourceQuery);
+
+  let publicPath = urlParts.publicPath;
+
+  // this tests if publicPath is relative, and makes it an absolute url if it is
+  // will match something of the form '/.../', or simply '/'
+  if (/^\/.*\/$|^\/$/.test(publicPath)) {
```
Comment (evilebottnawi) [inline client-src/default/utils/updatePublicPath.js]: Can we encapsulate this in `isRelativeUrl` (maybe using other repository for this check)?

File: client-src/default/index.js [hunk 2]
```diff
@@ -9,6 +9,9 @@ const { log, setLogLevel } = require('./utils/log');
 const sendMessage = require('./utils/sendMessage');
 const reloadApp = require('./utils/reloadApp');
 const createSocketUrl = require('./utils/createSocketUrl');
+const updatePublicPath = require('./utils/updatePublicPath');
+
+updatePublicPath(__resourceQuery);
```
Comment (evilebottnawi) [inline client-src/default/index.js]: I am afraid what we change global variable and it is can break code in some rare case, maybe we can rewrite this on using own variable like:
```
const resourceQuery = getPublicPath(__resourceQuery);
```

Comment (Loonride) [reply client-src/default/index.js]: Yes good point

**Comment by jony89** [comment]
Hey, thanks for the PR this is really needed. as the functionality for publicPath is broken now.

**Comment by evilebottnawi** [comment]
Only in next major release, sorry it is breaking change

**Comment by jony89** [comment]
> Only in next major release, sorry it is breaking change

sure but can this at least be merged for some of us that want to start work with the `next` branch?

**Comment by evilebottnawi** [comment]
@jony89 yep, we will do this in near future (now we are working on next release webpack-dev-middleware, so it will be updated too)

**Comment by jony89** [comment]
Hey, 21 days pass, any progress? can I help with anything?

**Pull request closed without merging**